library-best-practices/JLBP-1.md

```diff
@@ -0,0 +1,20 @@
+[JLBP-1] Minimize dependencies
+------------------------------
+
+Use the minimum number of dependencies that is reasonable.
+Adding a dependency for a large amount of functionality may be ok,
+but avoid pulling in dependencies just to save a few lines of code,
+because every dependency is a liability.
+
+Some specific notes about minimizing dependencies:
+
+- Use the smallest scope possible. For example, auto-value doesn't
+  need to use compile scope, and can instead use compile-only,
+  since it doesn't need to appear on the classpath of consumers.
+  - All libraries used only for testing should have test scope
+    (for example junit, mockito, and truth).
+- Scrutinize all dependency additions. Check the result of
+  `mvn dependency:tree` (after running `mvn install -DskipTests`
+  to build the library) to see which transitive dependencies are
+  added by just adding a single dependency to your own library,
+  and if you require all of the transitive dependencies.
```

Comment:
Is it worth expanding on "if you require all of the transitive dependencies"? Even if you exclude them in your project, any project that depends on yours will not see the "excludes" directives on transitive dependencies.

Reply:
I want to make sure I understand what you're saying.

Let's use the following example: say your library (C) depends on B which transitively depends on A, and in C you exclude A. It seems clear to me that if a user of C (say D) depends on C and also adds their own dependency on B, then they would pull in A (since they didn't add the exclusion of A). However, would D pull in A in the case where D *only* lists C as their dependency?

Reply:
Right. I believe that is true, because you don't want someone else to diasable a dependency that might be needed somewhere else. `<exclude>` is not persisted in a published artifacts pom.xml. But now I feel like I need to double check. Let me try it out.

Reply:
Sorry, I'm super wrong. I don't know why I thought this.

Reply:
Alrighty. I'm going to go ahead and merge as-is.